File: pr-env-api/src/services/dockerComposeService.js
const { logger } = require('../utils/logger');
const { executeCommand } = require('../utils/commandExecutor');
const fileSystem = require('../utils/fileSystem');
const {
  createEnvironmentId,
  getEnvironmentDir,
  updateEnvironmentFiles
} = require('../utils/environmentConfig');

/**
 * Copy and configure vertuoza-compose folder for PR environment
 * @param {string} repositoryName - Repository name
 * @param {number} prNumber - PR number
 * @param {Array} services - Array of services with name and image_url
 * @returns {Promise<string>} - Path to the configured folder
 */
async function setupPrEnvironment(repositoryName, prNumber, services) {
  try {
    // Create environment ID using repository name
    const environmentId = createEnvironmentId(repositoryName, prNumber);

    // Create environment directory
    const environmentDir = getEnvironmentDir(environmentId);
    await fileSystem.ensureDirectory(environmentDir);

    // Path to the source vertuoza-compose folder
    const sourceDir = fileSystem.joinPath(__dirname, '../../../vertuoza-compose');

    // Copy the entire vertuoza-compose folder contents directly to the environment directory
    await fileSystem.copyDirectory(sourceDir, environmentDir);

    // Update environment files with PR-specific configuration and multiple services
    await updateEnvironmentFiles(environmentDir, repositoryName, prNumber, services);

    logger.info(`Set up PR environment at ${environmentDir} with ${services.length} services`);

    return environmentDir;
  } catch (err) {
    logger.error(`Error setting up PR environment: ${err.message}`);
    throw err;
  }
}

/**
 * Start a Docker Compose environment
 * @param {string} environmentDir - Environment directory
 * @returns {Promise<void>}
 */
async function startEnvironment(environmentDir) {
  try {
    await executeCommand(`cd ${environmentDir} && docker compose up -d`);
    logger.info(`Started Docker Compose environment at ${environmentDir}`);
  } catch (err) {
    logger.error(`Error starting Docker Compose environment: ${err.message}`);
    throw err;
  }
}

/**
 * Stop and remove a Docker Compose environment
 * @param {string} environmentDir - Environment directory
 * @returns {Promise<void>}
 */
async function stopEnvironment(environmentDir) {
  try {
    await executeCommand(`cd ${environmentDir} && docker compose down -v --remove-orphans`);

    // Clean up any dangling containers, images, and volumes
    await executeCommand('docker system prune -f');

    logger.info(`Stopped Docker Compose environment at ${environmentDir}`);
  } catch (err) {
    logger.error(`Error stopping Docker Compose environment: ${err.message}`);
    throw err;
  }
}

/**
 * Clean up environment directory
 * @param {string} environmentDir - Environment directory
 * @returns {Promise<void>}
 */
async function cleanupEnvironment(environmentDir) {
  try {
    await fileSystem.remove(environmentDir);
    logger.info(`Cleaned up environment directory at ${environmentDir}`);
  } catch (err) {
    logger.error(`Error cleaning up environment directory: ${err.message}`);
    throw err;
  }
}

module.exports = {
  setupPrEnvironment,
  startEnvironment,
  stopEnvironment,
  cleanupEnvironment
};

File: pr-env-api-go/internal/services/dockercompose/dockercompose_service.go
package dockercompose

import (
	"fmt"
	"path/filepath"

	"github.com/vertuoza/pr-env-api/internal/models"
	"github.com/vertuoza/pr-env-api/internal/utils/commandexecutor"
	"github.com/vertuoza/pr-env-api/internal/utils/envconfig"
	"github.com/vertuoza/pr-env-api/internal/utils/filesystem"
	"github.com/vertuoza/pr-env-api/internal/utils/logger"
	"go.uber.org/zap"
)

// Service implements the DockerComposeService interface
type Service struct {
	executor *commandexecutor.CommandExecutor
	envConfig *envconfig.EnvironmentConfig
}

// NewService creates a new Docker Compose service
func NewService(executor *commandexecutor.CommandExecutor, envConfig *envconfig.EnvironmentConfig) *Service {
	return &Service{
		executor: executor,
		envConfig: envConfig,
	}
}

// SetupPrEnvironment sets up a PR environment with Docker Compose
func (s *Service) SetupPrEnvironment(repositoryName string, prNumber int, services []models.Service) (string, error) {
	// Create environment ID
	environmentID := s.envConfig.CreateEnvironmentID(repositoryName, prNumber)

	// Create environment directory
	environmentDir := s.envConfig.GetEnvironmentDir(environmentID)
	if err := filesystem.EnsureDirectory(environmentDir); err != nil {
		logger.Error("Failed to create environment directory", err,
			zap.String("directory", environmentDir))
		return "", err
	}

	// Path to the source vertuoza-compose folder
	sourceDir := filepath.Join("..", "vertuoza-compose")

	// Copy the entire vertuoza-compose folder contents directly to the environment directory
	if err := filesystem.CopyDirectory(sourceDir, environmentDir); err != nil {
		logger.Error("Failed to copy vertuoza-compose directory", err,
			zap.String("source", sourceDir),
			zap.String("destination", environmentDir))
		return "", err
	}

	// Update environment files with PR-specific configuration and multiple services
	if err := s.envConfig.UpdateEnvironmentFiles(environmentDir, repositoryName, prNumber, services); err != nil {
		logger.Error("Failed to update environment files", err,
			zap.String("directory", environmentDir))
		return "", err
	}

	logger.Info("Set up PR environment",
		zap.String("directory", environmentDir),
		zap.Int("services", len(services)))

	return environmentDir, nil
}

// StartEnvironment starts a Docker Compose environment
func (s *Service) StartEnvironment(environmentDir string) error {
	cmd := fmt.Sprintf("cd %s && docker compose up -d", environmentDir)
	_, err := s.executor.Execute(cmd)
	if err != nil {
		logger.Error("Failed to start Docker Compose environment", err,
			zap.String("directory", environmentDir))
		return err
	}

	logger.Info("Started Docker Compose environment", zap.String("directory", environmentDir))
	return nil
}

// StopEnvironment stops and removes a Docker Compose environment
func (s *Service) StopEnvironment(environmentDir string) error {
	cmd := fmt.Sprintf("cd %s && docker compose down -v --remove-orphans", environmentDir)
	_, err := s.executor.Execute(cmd)
	if err != nil {
		logger.Error("Failed to stop Docker Compose environment", err,
			zap.String("directory", environmentDir))
		return err
	}

	// Clean up any dangling containers, images, and volumes
	_, err = s.executor.Execute("docker system prune -f")
	if err != nil {
		logger.Error("Failed to prune Docker system", err)
		return err
	}

	logger.Info("Stopped Docker Compose environment", zap.String("directory", environmentDir))
	return nil
}

// CleanupEnvironment removes the environment directory
func (s *Service) CleanupEnvironment(environmentDir string) error {
	if err := filesystem.Remove(environmentDir); err != nil {
		logger.Error("Failed to clean up environment directory", err,
			zap.String("directory", environmentDir))
		return err
	}

	logger.Info("Cleaned up environment directory", zap.String("directory", environmentDir))
	return nil
}
